stag-library/src/main/java/com/vimeo/stag/UseStag.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * The MIT License (MIT)
+ * <p/>
+ * Copyright (c) 2016 Vimeo
+ * <p/>
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * <p/>
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ * <p/>
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.vimeo.stag;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Target;
+
+/**
+ * Use this annotation to tell Stag that all the member variables of the class should be processed.
+ * <p/>
+ * If the class is annotated, Stag will generate a TypeAdapter for that class.
+ * Stag does not generate TypeAdapters for abstract classes.
+ * <p/>
+ */
+@Target({ElementType.TYPE})
+public @interface UseStag {
```

Comment:
Thanks all for reviewing this. 

I will first begin by sharing the challenges when we were trying to move to Stag. We have more than 350+ model in different classes spread across different modules. Each of these classes have several member variables. We also do use @SerializedName in some of the cases. Putting a GsonAdapterKey to all these member variables was a cumbersome job. If the member variable name and the JSON serialized name was exactly the same then it was turning out an overkill for developers to add an annotation to every field. This is when we thought we add an annotation at a class level since STAG generates adapters and adapters are at a class level. So if one had to exclude a particular class they just do not add that Annotation to the class.

While raising this PR we did have the early adopters in mind, and hence we did retain the GsonAdapterKey for that backward compatibility. Existing STAG users will have to add an annotation at a class level to migrate to the never version - and hence the version bump up. 

As @kevinzetterstrom rightly mentions, keeping separate annotations for class level and field level was to have clarity in usage. Also like @anthonycr 's recommendation to add Options.ALL, Options.NONE we kept it separate as the class level specifications are always going to be different from the field level ones (Options.IGNORE_BASE_CLASS_FIELDS etc). It makes it more flexible in the future and we will not have to make breaking changes in future if we keep them separate.

To alleviate the pains of folks moving from v1 to v2 we can support GsonAdapterKey Or UseStag both - but that will end up in a code mess for people where half of the code is doing it in one way and the other half in a different way. Also for the library it will be an overhead to keep supporting both the ways.


@mcumings  : We will like to provide @Expose functionality at some point in time as well. When implemented the generate code will have an if-else condition which checks the field being exposed at run time. The if else clause will exist only for the fields which have @Expose annotation set. Also on the lines of what @anthonycr mentions we can have compile time exclusion policies in a future release.

If none of these works here is an alternative that should hopefully work for all - 

- Support GsonAdapterKey both at class and field level
- If this comes at the class level all member variables automatically are considered for parsing (irrespective of whether they have or do not have the GsonAdapterKey)
- For excluding only a few members from being serialized/deserialized do not include the GsonAdapterKey at the class level, but include it for the fields which need to be serialized/deserialized (not the ideal approach once again since it is only one or two fields which need to be excluded from the Serialization/Deserialization)